windows/src/engine/inst-cpp/insthelper.cpp

```diff
@@ -0,0 +1,211 @@
+#include "insthelper.h"
+#include "pch.h"
+#include <AccCtrl.h>
+#include <ShlObj_core.h>
+#include <iostream>
+#include <knownfolders.h>
+#include <limits.h>
+#include <msiquery.h>
+#include <security.h>
+#include <utility>
+#include <tchar.h>
+#include <string>
+#include <AclAPI.h>
+#include <Windows.h>
+#include <Msi.h>
+#include <initguid.h>
+#include <comdef.h>
+#include <msctf.h>
+#include <atlbase.h>
+#include<sstream>
+#include<iomanip>
+
+const LPCTSTR SFolderKeymanRoot = TEXT("\\Keyman");
+
+extern "C" __declspec(dllexport) unsigned int EnginePostInstall(MSIHANDLE hInstall) {
+  HANDLE hFile;
+
+  // Find %appdata% path
+  PWSTR path = new wchar_t[MAX_PATH];
+
+  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_ProgramData, 0, NULL, &path))) {
+    wcscat_s(path, MAX_PATH, SFolderKeymanRoot);
+
+    // Create directory if it does not exist
+    if (GetFileAttributes(path) == INVALID_FILE_ATTRIBUTES) {
+      if (!CreateDirectory(path, NULL)) {
+        DWORD errorCode = GetLastError();
+        std::wstring error =
+            L"Keyman Engine failed to set permissions on shared data in CreateDir: " + std::to_wstring(errorCode);
+        MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+        return errorCode;
+      }
+    }
+
+    // Create file handle
+    hFile = CreateFile(path, READ_CONTROL | WRITE_DAC, 0, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    if (hFile == INVALID_HANDLE_VALUE) {
+      DWORD errorCode    = GetLastError();
+      std::wstring error = L"Keyman Engine failed to set permissions on shared data in CreateFile: " + std::to_wstring(errorCode);
+      MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+      return errorCode;
+    }
+
+    // Set permission on shared data
+    EXPLICIT_ACCESS ea      = {0};
+    ea.grfAccessPermissions = GENERIC_READ | GENERIC_EXECUTE;
+    ea.grfAccessMode        = SET_ACCESS;
+    ea.grfInheritance       = SUB_CONTAINERS_AND_OBJECTS_INHERIT;
+    ea.Trustee.TrusteeForm  = TRUSTEE_IS_NAME;
+    ea.Trustee.ptstrName    = (LPWCH)L"ALL APPLICATION PACKAGES";
+
+    PACL pOldDACL = NULL;
+
+    DWORD result = SetEntriesInAcl(1, &ea, NULL, &pOldDACL);
+    if (result != ERROR_SUCCESS) {
+      CloseHandle(hFile);
+      std::wstring error =
+          L"Keyman Engine failed to set permissions on shared data in GrantPermission: " + std::to_wstring(result);
+      MsiSetProperty(hInstall, TEXT("EnginePostInstall_Error"), error.c_str());
+      if (pOldDACL) {
+        LocalFree(pOldDACL);
+      }
```

Comment:
There's a lot of cleanup in failure cases, and the number of cleanup items grows. A common pattern to handle this in C++ Windows code without SEH is `goto`, yes `goto`... e.g.

```c++
  DWORD result = ERROR_SUCCESS;
  HANDLE hFile = INVALID_HANDLE_VALUE;
  PACL pOldDACL = NULL;

  // ... extra code ...

  hFile = CreateFile(path, READ_CONTROL | WRITE_DAC, 0, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
  if (hFile == INVALID_HANDLE_VALUE) {
    result = GetLastError();
    ReportError(L"CreateFile", result);
    goto cleanup;
  }

  // ... extra code ...

  result = SetEntriesInAcl(1, &ea, NULL, &pOldDACL);
  if(result != ERROR_SUCCESS) {
    ReportError(L"SetEntriesInAcl", result);
    goto cleanup;
  }

  // ... extra code ...

  // eventually, everything passes, and if we get here than result == ERROR_SUCCESS already
cleanup:
  if(hFile != INVALID_HANDLE_VALUE) {
    CloseHandle(hFile);
  }
  if(pOldDACL != NULL) {
    LocalFree(pOldDACL);
  }
  return result;
}
```